src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -2372,12 +2402,71 @@ Node *PhiNode::Ideal(PhaseGVN *phase, bool can_reshape) {
       if (ii->is_MergeMem()) {
         MergeMemNode* n = ii->as_MergeMem();
         merge_width = MAX2(merge_width, n->req());
-        saw_self = saw_self || (n->base_memory() == this);
+        if (n->base_memory() == this) {
+          split_must_terminate = true;
+        }
       }
     }
 
-    // This restriction is temporarily necessary to ensure termination:
-    if (!saw_self && adr_type() == TypePtr::BOTTOM)  merge_width = 0;
+    // There are cases with circular dependencies between bottom Phis
+    // and MergeMems. Below is a minimal example.
+    //
+    //               +------------+
+    //               |            |
+    // (base_memory) v            |
+    //              MergeMem      |
+    //                 |          |
+    //                 v          |
+    //                Phi (this)  |
+    //                 |          |
+    //                 v          |
+    //                Phi         |
+    //                 |          |
+    //                 +----------+
+    //
+    // Here, we cannot break the circularity through a self-loop as there
+    // are two Phis involved. Repeatedly splitting the Phis through the
+    // MergeMem leads to non-termination. We check for non-termination below.
+    // Only check for non-termination if necessary.
+    if (!split_must_terminate && adr_type() == TypePtr::BOTTOM &&
+        merge_width > Compile::AliasIdxRaw) {
+      ResourceMark rm;
+      VectorSet visited;
+      Node_List worklist;
```

Comment:
You could also use a `Unique_Node_List` instead of a `Node_List` + `VectorSet`:
```
Unique_Node_List worklist;
for (uint i = 0; i < worklist.size(); i++) {
}
```

Reply:
Correct me if I'm wrong, but doesn't this pattern lead to higher memory consumption because we need to keep already visited nodes in the worklist? It probably doesn't matter much in practice, but I find the current approach more hygienic.

Reply:
That's true that we have more memory consumption. But it's probably neglectable. However, we seem to use both approaches throughout our code base, so I don't have a strong opinion about what which version we use here.

Reply:
OK, I'll leave it as is then for this changeset. Perhaps we should have a broader discussion at some point to make this consistent across a larger part of the code base.